fdas/main.go

```diff
@@ -0,0 +1,321 @@
+			return
+		}
+
+		err = db.Put([]byte(newACL.User+"/"+newACL.Object), []byte(newACL.Relation), nil)
+		if err != nil {
+			c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
+			return
+		}
+
+		c.IndentedJSON(http.StatusCreated, newACL)
+
+	}
+}
+
+func checkACL(tokens *Set, db *leveldb.DB, kv *capi.KV) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !tokens.Has(c.Param("token")) {
+			c.IndentedJSON(http.StatusUnauthorized, gin.H{"message": "unauthorized"})
+			return
+		}
+		object := c.Query("object")
+		relation := c.Query("relation")
+		user := c.Query("user")
+
+		if object == "" || relation == "" || user == "" {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL request"})
+			return
+		}
+
+		data, err := db.Get([]byte(user+"/"+object), nil)
+		if err != nil {
+			c.IndentedJSON(http.StatusNotFound, gin.H{"message": "ACL not found"})
+			return
+		}
+		if string(data) == relation {
+			c.IndentedJSON(http.StatusOK, true)
+			return
+		}
+		parts := strings.Split(object, ":")
+		if len(parts) != 2 {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL object definition"})
+			return
+		}
+		namespace := parts[0]
+		value, _, err := kv.Get(namespace+"/"+string(data)+"/"+relation, nil)
+		if err == nil || value != nil {
+			c.IndentedJSON(http.StatusOK, true)
+		} else {
+			c.IndentedJSON(http.StatusOK, false)
+		}
+
+	}
+}
+
+func postNamespace(tokens *Set, kv *capi.KV) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !tokens.Has(c.Param("token")) {
+			c.IndentedJSON(http.StatusUnauthorized, gin.H{"message": "unauthorized"})
+			return
+		}
+		var newNamespace fdasConfig
+		relatshionshipMap := make(map[string][]string)
+		allRelations := ""
+		if err := c.BindJSON(&newNamespace); err != nil {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid namespace definition"})
+			return
+		}
+		pair, _, err := kv.Get(newNamespace.Namespace, nil)
+		if err == nil || pair != nil {
+			keys, _, err := kv.Keys(newNamespace.Namespace, "", nil)
+			if err != nil {
+				c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
+				return
+			}
+			for _, key := range keys {
+				kv.Delete(key, nil)
+			}
+		}
+
+		for key, value := range newNamespace.Relations {
+			if allRelations == "" {
+				allRelations = key
+			} else {
+				allRelations += "," + key
+			}
+			relatshionshipMap[key] = value.Union
+
+			err = createRelations(newNamespace.Namespace, key, value.Union, &relatshionshipMap, kv)
+			if err != nil {
+				c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
+				return
+			}
+		}
+
+		p := &capi.KVPair{Key: newNamespace.Namespace, Value: []byte(allRelations)}
+		_, err = kv.Put(p, nil)
+		if err != nil {
+			c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
```

Comment:
Non-graceful fall

Reply:
- to return error message to user use err.msg
- or log it
- or panic it to log it to sys.err io stream

Reply:
I don't think that revealing to api user internal structure of our code and our service is appropriate way to handle errors. I will log it for our internal use, thank you for noticing